### ISSUE TITLE
How to turn Analytics on and off in code

### ISSUE THREAD

**Issue by newuser44** [arnesson/cordova-plugin-firebase#843]
Trying to figure out how to turn Analytics off and on for different environments.  
On for testing it in dev,  off for QA testing,  on for production.  

I see there is  setAnalyticsCollectionEnabled.  At first we didn't even call this and it was always on.   Maybe that is by default?   Even with this turned on, in the plist file there is value something like get Analytics, but this always false.  (and some ad value is always turned on).  
Maybe these are two seperate things.   

Does setAnalyticsCollectionEnabled turn analytics all the way off.  Even for first open and all that?

**Comment by newuser44** [comment]
Along with this is how people implement this now for different environments.   Looking at the Analytics doc, it seems the best approach is to have a different JSON (and plist) file for each environment or one for testing (dev and acceptance) and one for production.   That is where things seem to get complicated.  I have a file both android and ios,  so two firebase files (json and plist)  I would need to split that out for testing and production and get the plugin to know the correct file.  

How do you guys handle this now?

**Comment by soumak77** [comment]
There are two things that will be required:

1. adding variables to this plugin to configure analytics behavior
2. write a script which sets the value of the variables based on your environment by modifying `config.xml` before the build.

Cordova is responsible for generating the JSON and plist files, so you shouldn't try to modify those yourself.  When you set the value of a specific variable for a plugin, that value will make its way into those files.

I always have a debug and production firebase project for every app I create.  This ensures I don't pollute my production analytics, users, database, etc with test data.  I have custom scripts which copies the correct firebase config files based on the environment.  I also modify my `config.xml` to set the variables for various plugins.

For example, I use `cordova-plugin-facebook4` which requires settings variables `APP_ID` and `APP_NAME`.  I have separate apps in facebook for debug and production, so I must modify `config.xml` to set the correct values based on whether I'm testing my app or releasing a new production version.

**Comment by newuser44** [comment]
I agree it seems like have to have different files.  The tricky part is tell the build which file to use.

**Comment by briantq** [comment]
@newuser44 Since the config files have my API keys and other sensitive data, I don't check in my production files.  Instead, I do check in my dev profiles that allow developers to easily build locally and test using the provided infrastructure. 

My build is set up in such a way that I overwrite the dev version of the files with the production version of the files prior to building and deploying.  This allows me to keep the production keys/files secure, allow devs to test on the dev environment without any changes, and simplify the code as it always loads the same file. 

It seems to check all the best practices boxes and is pretty straight forward to set up in Jenkins.

**Comment by soumak77** [comment]
Yeah, I've had to spend a lot of time developing custom scripts to get the builds to copy/config my apps based on my desired environment.  I have 3 specific build types `local`, `debug`, and `release`.  Files I need for each environment are contained in specific folders: `environments/local`, `environments/debug` and `environments/release`.  Each folder contains the specific JSON and plist files from firebase, the firebase web config files, as well as a `config.json` which contains specific values I need to set for various plugins.

I then have custom scripts which accept an argument specifying the build type.  The script then grabs the files from the desired environment folder and copies the files to the correct location.  The script also copies the values from `config.json` and puts them into `config.xml` for each plugin.

**Comment by soumak77** [comment]
I agree with @briantq, never check in production config files into source control, only debug files.

**Comment by soumak77** [comment]
@newuser44 as a curtesy to you, as well as others in the community, I've open sourced the build scripts I use for my cordova/ionic apps.  You can check them out here: https://github.com/soumak77/build-scripts.  Hopefully that gives you a starting point for how to customize your builds based on your environment.

**Comment by newuser44** [comment]
We do have a webpack.config.js script that I use to build.  We have environment FILES.  environment.accept.   environment.dev  and so forth.     I was just going to ask if you could share your scripts. Looks like you just did. Thanks

**Comment by newuser44** [comment]
I have been updating our webpack.config to setup firebase for testing and production. 
I got it working on the android side and shifted over to apple.  It seems like apple is more tied to the Bundle Identifier to the point that you have to have that setup in some certificate?   For android I was able to just do com.appname.cool.test   as the app id and setup firebase.  

Do you have to set up the app twice on apple to handle two different bundle identifiers one for debug or testing and one from production?

**Comment by soumak77** [comment]
I know @brianqt has been working heavily on iOS development with this plugin. Perhaps he can provide more details on how to bypass the signing requirement for debug builds for an iOS app.

**Comment by briantq** [comment]
@newuser44 @soumak77 I have not used Google Analytics so I don't know what the requirements are.  

In general, if you need two different apps ids, then yes, you will need to set up two different versions of the plugin.  I needed to do that for testing Firebase push in my application.  I set up two branches with the different build ids to handle it.

**Comment by briantq** [comment]
@newuser44 sorry, I am not an expert with provisioning profiles and such.  You might try online as I won't be much help.

Also, are you still having issues with the original problem of enabling/disabling Analytics?  It seems that the discussion has gotten off the original topic.  That is ok, but wanted to see if the original issue still exists.

**Comment by newuser44** [comment]
We have not been using that functionality to turn it on.  It seems like Analytics is on by default.  I have not tried to turn it off.    Maybe after I have got this split up into testing and production analytics and I try it. 
(I guess there is also a way to turn performance on and off)
